src/com/xilinx/rapidwright/eco/ECOPlacementHelper.java

```diff
@@ -0,0 +1,344 @@
+                int isUpper = pairID > 'D' ? 1 : 0;
+
+                SitePinInst existingClkSpi = siteInst.getSitePinInst(clkSitePinNames[isUpper]);
+                Net existingClk = existingClkSpi != null ? existingClkSpi.getNet() : null;
+                if (existingClk != null && existingClk != clk) {
+                    // Allow pre-existing SitePinInsts if they were deferred for removal
+                    if (deferredRemovals != null && !deferredRemovals.getOrDefault(existingClk, Collections.emptySet()).contains(existingClkSpi)) {
+                        continue;
+                    }
+                }
+
+                // Check that CE and SR are VCC and GND respectively
+                Pair<String, String> p = belTypeSitePinNameMapping.get(belFlop);
+                Net existingCE = siteInst.getNetFromSiteWire(p.getFirst());
+                if (existingCE != null && existingCE.getType() != NetType.VCC) {
+                    continue;
+                }
+                Net existingSR = siteInst.getNetFromSiteWire(p.getSecond());
+                if (existingSR != null && existingSR.getType() != NetType.GND) {
+                    continue;
+                }
+
+                // Compatible BEL found! Return.
+                return bel;
+            }
+        }
+
+        // This site has no compatible flops -- remember it for next time
+        if (flopLessSites == null) {
+            flopLessSites = new HashSet<>();
+            flopLessSitesByClk.put(clk, flopLessSites);
+        }
+        flopLessSites.add(site);
+        return null;
+    }
+
+    /**
+     * Given a SiteInst, find an unused LUT BEL that can host a new LUT6 cell.
+     *
+     * @param siteInst SiteInst object to search inside.
+     * @return Unused LUT6 BEL.
+     */
+    public BEL getUnusedLUT(SiteInst siteInst) {
+        Site site = siteInst.getSite();
+        if (lutLessSites.contains(site)) return null;
+        if (!siteInst.getName().startsWith(SiteInst.STATIC_SOURCE)) {
+            for (Character belLUT : LUTTools.lutLetters) {
+                // Check both LUTs are unoccupied, try something fancy later (TODO)
+                String lut6Name = belLUT + "6LUT";
+                Cell lut6 = siteInst.getCell(lut6Name);
+                Cell lut5 = siteInst.getCell(belLUT + "5LUT");
+                if (lut6 != null || lut5 != null) continue;
+
+                // Check if LUT is supplying GND/VCC
+                String lutOutput = belLUT + "_O";
+                SitePinInst pinInst = siteInst.getSitePinInst(belLUT + "_O");
+                if (pinInst != null) continue;
+                if (siteInst.getNetFromSiteWire(lutOutput) != null) continue;
+
+                // Assume not being used as a thru-site PIP (TODO)
+                return siteInst.getBEL(lut6Name);
+            }
+        }
+
+        lutLessSites.add(site);
+        return null;
+    }
+
+    private Site getSite(Site site, int dx, int dy) {
```

Comment:
I can't remember if I copied this pattern from somewhere or I preemptively assumed it would be too slow, but the advantage of having a static method does seem quite useful so I've removed this lookup now